### DISCUSSION TITLE
Why does `colour.RGB_to_XYZ` definition output differs from Bruce Lindbloom's calculator?

### DISCUSSION THREAD

**Discussion by blue-j** [colour-science/colour#1125] [Q&A] [answered]
I have a script that does a Mini Batch Kmeans clustering with sklearn and then calculates the XYZ value of each RGB cluster.  Unfortunately, I cant find anywhere where the output can be sanctified.  [Bruce Lindbloom's calculator](http://www.brucelindbloom.com/index.html?ColorCalculator.html) says I am wrong... and so I come here to see, have I made a mistake in the code?

```
    # Set variables to be used to convert to CIEXYZ
    wt = np.array(list(map(float, args.whitexy.split(','))))
    rd = np.array(list(map(float, args.redxy.split(','))))
    gr = np.array(list(map(float, args.greenxy.split(','))))
    bl = np.array(list(map(float, args.bluexy.split(','))))

    # this is D65; if D50 is desired, use 0.3457, 0.3585 for illuminant_XYZ
    illuminant_XYZ = np.array([0.3128, 0.3289])
    illuminant_RGB = wt
    chromatic_adaptation_transform = "Bradford"
    p = np.array([rd, gr, bl])
    matrix_RGB_to_XYZ = colour.normalised_primary_matrix(p, wt)

    # Cluster the input image
    clustered_image, labels, kmeans = cluster_image(args.input_image, args.num_clusters)
    total_pixels = labels.size
    res = []
    for i, color in enumerate(kmeans.cluster_centers_):
        rgb = np.array([color[0] / 255, color[1] / 255, color[2] / 255])
        hsv = colour.RGB_to_HSV(rgb)

        # if white points are equal, we do not use chromatic adaptation
        if np.allclose(wt, illuminant_XYZ):
            ciexyz = colour.RGB_to_XYZ(
                rgb,
                illuminant_RGB,
                illuminant_XYZ,
                matrix_RGB_to_XYZ,
            )
        else:
            ciexyz = colour.RGB_to_XYZ(
                rgb,
                illuminant_RGB,
                illuminant_XYZ,
                matrix_RGB_to_XYZ,
                chromatic_adaptation_transform,
            )

```
I am not getting any errors and the code looks good to me (it's much longer, but I have spared you!) but what i am concerned about is my execution of the RGB to XYZ function.  sRGB is the source space. 

this is what the args look like from the command line for an sRGB image:
"0.3128, 0.3289" "0.6488, 0.3304" "0.3211, 0.598" "0.1558, 0.0664"

for RGB (82, 115, 169) i get an XYZ of:    0.42641328  0.44800904  0.7002739

apparently, this is not correct?   RGB (158, 180, 214) is the color reported from those XYZ coordinates (0.42641328  0.44800904  0.7002739) in Bruce's calculator.

I also tried of course to enter the RGB (82, 115, 169) with no adaptation, and Bruce's XYZ is: 0.167694 0.169187 0.399104

Thank you in advance for any time you may take to review this.  : )   -J

**Comment by blue-j** [comment, marked as answer]
This is what the NPM looks like for this example:

[[ 0.36185485  0.39314477  0.19604934]
 [ 0.1842738   0.73217244  0.08355376]
 [ 0.01160077  0.09905142  0.97873668]]

**Reply by thomasmansencal** [reply]
Hey @blue-j, I’m at GDC, so very little bandwidth but I will look at that in a few days!

**Reply by thomasmansencal** [reply]
Hi,

So Lindbloom is correct:

```python
In [1]: import colour

In [2]: colour.sRGB_to_XYZ([82 / 255, 115 / 255, 169 / 255])
Out[2]: array([ 0.16771841,  0.16919879,  0.39918009])
```

What you need to do is pass the decoding CCTF in your calls, e.g.:


```python
cctf_decoding = colour.models.cctf_decoding_sRGB

...
            ciexyz = colour.RGB_to_XYZ(
                rgb,
                illuminant_RGB,
                illuminant_XYZ,
                matrix_RGB_to_XYZ,
                cctf_decoding=colour.models.cctf_decoding_sRGB
            )
```

**Reply by thomasmansencal** [reply]
Note that I'm thinking about making the `colour.RGB_to_XYZ` and `colour.XYZ_to_RGB` definitions a bit easier to use: https://github.com/colour-science/colour/issues/1127

**Reply by blue-j** [reply]
Interesting!  I have not heard of CCTF encoding before, so it's very educational for me.  Thank you again for your generous spirit!

I do see an issue with my use case, and wonder how to navigate this.  Colour supports only a few RGB spaces in the CCTF encodings, notably missing Adobe RGB, Display P3, and many more.  What route do you suggest?  I am converting the input colors into 70+ spaces - all RGB spaces supported by Colour, and most of the CIE ones.  I guess I could go to CIELAB, then to XYZ?  I was going to XYZ before converting to all the other spaces.  Thanks again!  - J

**Reply by blue-j** [reply]
It seems I can calculate the CCTF if I know the primaries and white point in xy?  this seems like the road to joy:

http://www.brucelindbloom.com/Eqn_RGB_XYZ_Matrix.html

Maybe you could incorporate these matrices that are missing from Colour?   I may be playing the fool, forgive me, I am an amateur.  - J

**Reply by thomasmansencal** [reply]
> Colour supports only a few RGB spaces in the CCTF encodings, notably missing Adobe RGB, Display P3, and many more. What route do you suggest? I am converting the input colors into 70+ spaces - all RGB spaces supported by Colour

A single colour component transfer function (CCT) can be used by many RGB colourspaces. But each RGB colourspaces has two, one for encoding one for decoding:

```python
In [1]: import colour

In [2]: colour.models.RGB_COLOURSPACE_DISPLAY_P3
Out[2]:
RGB_Colourspace('Display P3',
                [[ 0.68 ,  0.32 ],
                 [ 0.265,  0.69 ],
                 [ 0.15 ,  0.06 ]],
                [ 0.3127,  0.329 ],
                'D65',
                [[  4.86570949e-01,   2.65667693e-01,   1.98217285e-01],
                 [  2.28974564e-01,   6.91738522e-01,   7.92869141e-02],
                 [ -3.97207552e-17,   4.51133819e-02,   1.04394437e+00]],
                [[ 2.49349691, -0.93138362, -0.40271078],
                 [-0.82948897,  1.76266406,  0.02362469],
                 [ 0.03584583, -0.07617239,  0.95688452]],
                eotf_inverse_sRGB,
                eotf_sRGB,
                False,
                False)
```

> It seems I can calculate the CCTF if I know the primaries and white point in xy?

No, you can compute the normalised primary matrix (NPM) with them.

> Maybe you could incorporate these matrices that are missing from Colour?

We have them all already!

**Reply by blue-j** [reply]
i hope i do not try your patience by asking what probably is a very basic question, but when do you apply the CCTF / EOTFs?  how do you know when to use them?  encoding, decoding... which?

**Reply by thomasmansencal** [reply]
See [Poynton's Gamma Faq](https://poynton.ca/notes/colour_and_gamma/GammaFAQ.html), they are "Gamma Correction".

**Reply by blue-j** [reply]
Thank you for the resource!  : )

-J
